### PR TITLE
feat: warn against named parent routes

### LINKED ISSUE #1376: Warn against named parent routes
From https://github.com/vuejs/router/pull/1339

A route with a name and a child with an empty path without a name should warn when adding the route (inside of the matcher)

```js
const routes = [
  // this route should warn
  {
    name: 'UserRoute',
    path: '/user/:id',
    component: User,
    children: [
      { path: '', component: UserHome },
    ],
  },

// this shouldn't
  {
    name: 'UserRoute',
    path: '/user/:id',
    component: User,
    children: [
      { path: '', name: 'UserHome` component: UserHome },
    ],
  },
]
```

### PR DESCRIPTION
**Pull request by Eating-Eating**
(fix #1376[,#1376])
A route with a name and a child with an empty path without a name should warn when adding the route (inside of the matcher)

### CODE REVIEW

File: src/matcher/index.ts [hunk 1]
```diff
@@ -79,6 +79,8 @@ export function createRouterMatcher(
     // used later on to remove by name
     const isRootAdd = !originalRecord
     const mainNormalizedRecord = normalizeRouteRecord(record)
+    if (__DEV__)
+      checkChildMissingNameWithEmptyPath(mainNormalizedRecord, parent)
```
Comment (posva) [inline src/matcher/index.ts:83]:
```suggestion
    if (__DEV__) {
      checkChildMissingNameWithEmptyPath(mainNormalizedRecord, parent)
    }
```

**Review by posva — CHANGES_REQUESTED** [review]
Thanks! Can you also make sure this work with multiple levels of nested routes with empty paths?

File: __tests__/matcher/addingRemoving.spec.ts [hunk 2]
```diff
@@ -441,5 +441,33 @@ describe('Matcher: adding and removing records', () => {
       )
       expect('same param named').not.toHaveBeenWarned()
     })
+    it('warn if child route missing a required name', () => {
```
[the remark this answers is not in the document]
Comment (Eating-Eating) [reply __tests__/matcher/addingRemoving.spec.ts:444]: Thanks for your reply, should I add another test case or just change this case into a multiple levels of nested routes?

**Comment by Eating-Eating** [comment]
> Thanks! Can you also make sure this work with multiple levels of nested routes with empty paths?

Thanks for your reply, should I add another test case or just change this case into a multiple levels of nested routes?

**Comment by posva** [comment]
Adding a new test case and ensuring the implementation works 🙂

File: src/matcher/index.ts [hunk 3]
```diff
@@ -451,6 +454,28 @@ function checkSameParams(a: RouteRecordMatcher, b: RouteRecordMatcher) {
...
+  parent?: RouteRecordMatcher
+) {
+  if (
+    parent &&
+    parent.record.name &&
+    !mainNormalizedRecord.name &&
+    !mainNormalizedRecord.path
+  ) {
+    warn(
+      `A child route with an empth path that has a named parent route needs a name`
```
Comment (posva) [inline src/matcher/index.ts:474]:
```suggestion
      `The route named "${parent.record.name}" has a child without a name and an empty path. Using that name won't render the empty path child so you probably want to move the name to the child instead. If this is intentional, add a name to the child route to remove the warning.
```

File: __tests__/matcher/addingRemoving.spec.ts [hunk 4]
```diff
@@ -441,5 +441,79 @@ describe('Matcher: adding and removing records', () => {
...
+          {
+            name: 'UserRoute',
+            path: '/user/:id',
+            component: component,
+            children: [{ path: '', component: component }],
+          },
+        ],
+        {}
+      )
+      expect('route needs a name').toHaveBeenWarned()
```
Comment (posva) [inline __tests__/matcher/addingRemoving.spec.ts:457]:
```suggestion
      expect('has a child without a name').toHaveBeenWarned()
```

File: __tests__/matcher/addingRemoving.spec.ts [hunk 5]
```diff
@@ -441,5 +441,79 @@ describe('Matcher: adding and removing records', () => {
...
+          {
+            name: 'UserRoute',
+            path: '/user/:id',
+            component: component,
+            children: [{ path: '', name: 'UserHome', component: component }],
+          },
+        ],
+        {}
+      )
+      expect('route needs a name').not.toHaveBeenWarned()
```
Comment (posva) [inline __tests__/matcher/addingRemoving.spec.ts:472]:
```suggestion
      expect('has a child without a name').not.toHaveBeenWarned()
```

File: __tests__/matcher/addingRemoving.spec.ts [hunk 6]
```diff
@@ -441,5 +441,79 @@ describe('Matcher: adding and removing records', () => {
...
+                name: 'BBBB',
+                component: component,
+                children: [{ path: '', component: component }],
+              },
+            ],
+          },
+        ],
+        {}
+      )
+      expect('route needs a name').toHaveBeenWarned()
```
Comment (posva) [inline __tests__/matcher/addingRemoving.spec.ts:494]:
```suggestion
      expect('has a child without a name').toHaveBeenWarned()
```

File: __tests__/matcher/addingRemoving.spec.ts [hunk 7]
```diff
@@ -441,5 +441,79 @@ describe('Matcher: adding and removing records', () => {
...
+                name: 'BBBB',
+                component: component,
+                children: [{ path: '', name: 'CCCC', component: component }],
+              },
+            ],
+          },
+        ],
+        {}
+      )
+      expect('route needs a name').not.toHaveBeenWarned()
```
Comment (posva) [inline __tests__/matcher/addingRemoving.spec.ts:516]:
```suggestion
      expect('has a child without a name').not.toHaveBeenWarned()
```

**Comment by posva** [comment]
Thanks, I did some refactor on my end but didn't have push rights: https://github.com/vuejs/router/pull/1396

**Pull request closed without merging**